### ISSUE TITLE
Mask2Former model wrapper

### ISSUE THREAD

**Issue by JLrumberger** [Kainmueller-Lab/Pathology-Foundation-Model-Benchmark#8]
### Instructions
Please compile a design doc for the  [mask2former](https://github.com/facebookresearch/dinov2/blob/e1277af2ba9496fbadf7aec6eba56e8d882d1e35/dinov2/eval/segmentation_m2f/models/decode_heads/mask2former_head.py) + [vit adapter](https://github.com/facebookresearch/dinov2/blob/e1277af2ba9496fbadf7aec6eba56e8d882d1e35/dinov2/eval/segmentation_m2f/models/backbones/vit_adapter.py#L20) model wrapper class

Write a model wrapper that can be used for each of the foundation model backbones. The simple version should 
1. Initialize the different foundation models and their preprocessing function based on a name string
2. Add a trainable m2former (backbone, pixel decoder, and Transformer decoder) on top for dense prediction tasks.

**Design overview**
1. `M2FSegmentationModel` class with the foundation model encoder as `self.model` and the associated preprocessing as `self.transform`. The M2F is added as the head of the model. We basically create a wrapper around both the dinov2 backbone and the M2F segmented head, instantiate them, and make a forward pass to get our segmentation logits.

**Code mockup**

```python
import torch
import math
import itertools
import mmcv
import torch
import timm
import mmseg 

from functools import partial
import torch.nn.functional as F
from mmcv import runner
from torchmetrics.classification import MulticlassJaccardIndex

import dinov2.eval.segmentation.models

class CenterPadding(torch.nn.Module):
    def __init__(self, multiple):
        super().__init__()
        self.multiple = multiple

    def _get_pad(self, size):
        new_size = math.ceil(size / self.multiple) * self.multiple
        pad_size = new_size - size
        pad_size_left = pad_size // 2
        pad_size_right = pad_size - pad_size_left
        return pad_size_left, pad_size_right

    @torch.inference_mode()
    def forward(self, x):
        pads = list(itertools.chain.from_iterable(self._get_pad(m) for m in x.shape[:1:-1]))
        output = F.pad(x, pads)
        return output

class M2FSegmentationModel(torch.nn.Module):
    def __init__(self, model_name, num_classes,
               head_config_url, head_checkpoint_url,
               cfg_str,
               head_type="ms", head_scale_count=3,
               ):
         self.head_type = head_type
         self.head_scale_count = head_scale_count
        
        # Ensure that cfg_str is a valid path or content for a configuration file
        cfg = mmcv.Config.fromstring(cfg_str, file_format=".py")
        if self.head_type == "ms":
             cfg.data.test.pipeline[1]["img_ratios"] = cfg.data.test.pipeline[1]["img_ratios"][:self.head_scale_count]
             print("scales:", cfg.data.test.pipeline[1]["img_ratios"])
    
        self.backbone, self.transform = self.load_model_and_transform(model_name)
        self.backbone.eval()
        self.backbone.cuda()

        self.model = self.create_segmenter(cfg, backbone_model=backbone_model)
        runner.load_checkpoint(self.model, head_checkpoint_url, map_location="CPU")
        self.model.cuda()
        self.model.eval()

    def create_backbone(self, backbone_size="small"):
         # backbone_size in ("small", "base", "large" or "giant")
         BACKBONE_ARCHS = {
            "small": "vits14",
            "base": "vitb14",
            "large": "vitl14",
            "giant": "vitg14",
        }
        backbone_arch = BACKBONE_ARCHS[backbone_size]
        backbone_name = f"dinov2_{backbone_arch}"

        backbone_model = torch.hub.load(repo_or_dir="facebookresearch/dinov2", model=backbone_name)
        return backbone_model

    def create_segmenter(self, cfg, backbone_model):
        model = mmseg.apis.init_segmentor(cfg)
        model.backbone.forward = partial(
            backbone_model.get_intermediate_layers,
            n=cfg.model.backbone.out_indices,
            reshape=True,
        )
        if hasattr(backbone_model, "patch_size"):
            model.backbone.register_forward_pre_hook(lambda _, x: CenterPadding(backbone_model.patch_size)(x[0]))
        model.init_weights()
        return model

    def forward(self, x):
        b, c, h, w = x.shape
        x = self.transform(x)
       
        in_array = np.array(x)[:, :, ::-1] # BGR
        segmentation_logits = mmseg.apis.inference_segmentor(self.model, in_array)[0]
        return segmentation_logits

    def load_model_and_transform(self, model_name):
        if model_name == "UNI":
            model = timm.create_model("hf-hub:MahmoodLab/UNI", pretrained=True, init_values=1e-5, dynamic_img_size=True)
            transform = bio_image_datasets.create_transform(**resolve_data_config(model.pretrained_cfg, model=model))
        elif model_name == "Schuerch":
            raise NotImplementedError(f"Model {model_name} is not implemented.")
        elif model_name == "Virchow":
            raise NotImplementedError(f"Model {model_name} is not implemented.")
        else:
            raise NotImplementedError(f"Model {model_name} is not implemented.")
        return model, transform


def render_segmentation(segmentation_logits, dataset):
        colormap = DATASET_COLORMAPS[dataset]
        colormap_array = np.array(colormap, dtype=np.uint8)
        segmentation_values = colormap_array[segmentation_logits + 1]
        return Image.fromarray(segmentation_values)

if __name__ == "main":
    import numpy as np
    import dinov2.eval.segmentation.utils.colormaps as colormaps
    DATASET_COLORMAPS = {
        "ade20k": colormaps.ADE20K_COLORMAP,
        "voc2012": colormaps.VOC2012_COLORMAP,
    }

    model = M2FSegmentationModel(
               model_name='UNI', num_classes=8,
               head_config_url=???, head_checkpoint_url=???,
               cfg_str=???, head_type="ms", head_scale_count=3
    )
    dataset = LizardDataset(...)
    sample = dataset(111)
    image = sample['image']
    
    segmentation_logits = model(image)
    segmented_image = utils.render_segmentation(segmentation_logits, HEAD_DATASET)
    display(segmented_image)
    
    # Assuming `segmentation_logits` is the predicted segmentation map and `ground_truth` is the actual label
    score = MulticlassJaccardIndex(num_classes=8, average='macro')
    score_value = score(segmentation_logits, ground_truth)
    print("Jaccard Score:", score_value)
```

**Required inputs**
config string for head, foundation model checkpoints, and preprocessing functions.

**Output files**
None

**Timeline**
- 1-2 days

Estimated date when a fully implemented version will be ready for review:
Friday, 6th Dec 24

Estimated date when the finalized project will be merged in:
Friday, 6th Dec 24
